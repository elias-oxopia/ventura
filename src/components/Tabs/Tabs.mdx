---
name: Tabs
menu: Components
route: /tabs
---

import { Playground, Props } from 'docz';
import { Tabs, Tab } from '../..';

# Tabs

Tabs organize and allow navigation between groups of content that are related and at the same level of hierarchy.

## Examples

### Basic usage

<Playground>
  {() => {
    const [currentTabIndex, setCurrentTabIndex] = React.useState(0)
    return (
      <Tabs currentTabIndex={currentTabIndex} onChange={(tabIndex) => setCurrentTabIndex(tabIndex)}>
        <Tab title="First" index={0}>This is the first tab</Tab>
        <Tab title="Second" index={1}>This is the second tab</Tab>
        <Tab title="Third" index={2}>This is the third tab</Tab>
      </Tabs>
    )
  }}
</Playground>

## API

<Props of={Tabs} />
<Props of={Tab} />
