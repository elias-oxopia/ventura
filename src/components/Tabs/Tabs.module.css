.tabs {
}

.header {
  padding: 10px 0;
  border-bottom: 2px solid var(--color-neutral-2);
  margin-bottom: 25px;
}

.tabName {
  border: none;
  padding: 9px 14px;
  border-radius: var(--border-radius-small);
  font-size: var(--font-size);
  color: var(--color-neutral-7);
  background-color: var(--color-neutral-1);
  cursor: pointer;
}

.tabName:hover {
  background-color: var(--color-neutral-2);
  transition-duration: 0.2s;
}

.tabName.active {
  color: var(--color-neutral-8);
  font-weight: 600;
  background-color: var(--color-neutral-2);
  cursor: initial;
}

.tabName + .tabName {
  margin-left: 5px;
}
