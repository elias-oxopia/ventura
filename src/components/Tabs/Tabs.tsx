import React, { ReactNode, useState } from 'react';
import classNames from 'classnames';

import cssReset from '../../css-reset.module.css';
import styles from './Tabs.module.css';

interface TabsContextValue {
  currentTabIndex: number;
  addTab: (title: string, index: number) => void;
  removeTab: (index: number) => void;
}

export const TabsContext = React.createContext<TabsContextValue | null>(null);

interface Props {
  currentTabIndex: number;
  onChange: (tabIndex: number) => void;
  className?: string;
  children?: ReactNode;
}

const Tabs: React.FC<Props> = ({ currentTabIndex, onChange, className, children }: Props) => {
  const [tabNames, setTabNames] = useState<string[]>([]);
  const mergedClassNames = classNames(cssReset.ventura, styles.tabs, className);

  const onTabNameClick = (tabIndex: number) => {
    if (onChange) {
      onChange(tabIndex);
    }
  };

  const addTab = (title: string, index: number) => {
    setTabNames((previousTabNames) => {
      const newTabNames = [...previousTabNames];
      newTabNames[index] = title;
      return newTabNames;
    });
  };

  const removeTab = (index: number) => {
    setTabNames((previousTabNames) => previousTabNames.splice(index, 1));
  };

  const tabsContextValues: TabsContextValue = {
    currentTabIndex,
    addTab,
    removeTab,
  };

  return (
    <TabsContext.Provider value={tabsContextValues}>
      <div className={mergedClassNames}>
        <div className={styles.header}>
          {tabNames.map((text, index) => (
            <button
              key={text}
              type="button"
              className={classNames(cssReset.ventura, styles.tabName, {
                [styles.active]: currentTabIndex === index,
              })}
              onClick={() => onTabNameClick(index)}
            >
              {text}
            </button>
          ))}
        </div>
      </div>
      {children}
    </TabsContext.Provider>
  );
};

export default Tabs;
