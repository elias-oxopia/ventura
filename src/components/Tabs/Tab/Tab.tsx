import React, { ReactNode, useContext, useEffect } from 'react';
import classNames from 'classnames';

import { TabsContext } from '../Tabs';

import cssReset from '../../../css-reset.module.css';

export interface Props {
  title: string;
  index: number;
  className?: string;
  children: ReactNode;
}

const Tab: React.FC<Props> = ({ title, index, className, children }: Props) => {
  const tabsContext = useContext(TabsContext);

  useEffect(() => {
    if (!tabsContext) return undefined;

    tabsContext.addTab(title, index);

    return () => {
      tabsContext.removeTab(index);
    };
    // adding tabsContext causes infinite loop of updates
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [title, index]);

  if (tabsContext === null) {
    // eslint-disable-next-line no-console
    console.warn(`The ${title} Tab component needs to wrapped in a Tabs component.`);
    return null;
  }

  const mergedClassNames = classNames(cssReset.ventura, className);

  return tabsContext.currentTabIndex === index ? (
    <div className={mergedClassNames}>{children}</div>
  ) : null;
};

export default Tab;
