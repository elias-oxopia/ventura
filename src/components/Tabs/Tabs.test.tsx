import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import Tab from './Tab/Tab';
import Tabs from './Tabs';

describe('Tabs', () => {
  const defaultProps = {
    currentTabIndex: 0,
    onChange: jest.fn(),
  };

  const createComponent = (props: Omit<React.ComponentProps<typeof Tabs>, 'children'>) => (
    <Tabs {...props}>
      <Tab title="First" index={0}>
        This is the first tab
      </Tab>
      <Tab title="Second" index={1}>
        This is the second tab
      </Tab>
      <Tab title="Third" index={2}>
        This is the third tab
      </Tab>
    </Tabs>
  );

  test('default snapshot', () => {
    const { asFragment } = render(createComponent(defaultProps));
    expect(asFragment()).toMatchSnapshot();
  });

  test('onChange should trigger when another tab is clicked', () => {
    const { getByText } = render(createComponent(defaultProps));
    const secondTab = getByText(/Second/);
    fireEvent.click(secondTab);
    expect(defaultProps.onChange).toHaveBeenNthCalledWith(1, 1);
  });

  test('Second tab should be visible', () => {
    const { queryByText } = render(createComponent({ ...defaultProps, currentTabIndex: 1 }));
    const firstTab = queryByText(/This is the first tab/);
    const secondTab = queryByText(/This is the second tab/);
    expect(firstTab).toBe(null);
    expect(secondTab).not.toBe(null);
  });

  test('className prop', () => {
    const className = 'center';
    const { getByText } = render(createComponent({ ...defaultProps, className }));
    const tabsElement = getByText(/First/).parentElement?.parentElement!;

    const renderedClassNames = tabsElement.className.split(' ');
    expect(renderedClassNames).toContain(className);
    // className in prop should be the last in the row
    expect(renderedClassNames.indexOf(className)).toBe(renderedClassNames.length - 1);
  });
});
