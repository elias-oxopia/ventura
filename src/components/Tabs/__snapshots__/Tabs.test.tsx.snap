// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input default snapshot 1`] = `
<DocumentFragment>
  <div
    class="ventura tabs"
  >
    <div
      class="header"
    >
      <button
        class="ventura tabName active"
        type="button"
      >
        First
      </button>
      <button
        class="ventura tabName"
        type="button"
      >
        Second
      </button>
      <button
        class="ventura tabName"
        type="button"
      >
        Third
      </button>
    </div>
  </div>
  <div
    class="ventura"
  >
    This is the first tab
  </div>
</DocumentFragment>
`;
